File: csharp/patterns/src/PrototypePatterns/Behavioral/Mediator/MediatorPattern.cs
namespace PrototypePatterns.Behavioral.Mediator;

public sealed class ChatMediator
{
    private readonly List<Participant> _participants = [];

    public void Register(Participant participant) => _participants.Add(participant);

    public void Broadcast(Participant sender, string message)
    {
        foreach (var participant in _participants.Where(participant => participant != sender))
        {
            participant.Receive(message);
        }
    }
}

public sealed class Participant
{
    private readonly ChatMediator _mediator;

    public Participant(string name, ChatMediator mediator)
    {
        Name = name;
        _mediator = mediator;
        _mediator.Register(this);
    }

    public string Name { get; }

    public List<string> Inbox { get; } = [];

    public void Send(string message) => _mediator.Broadcast(this, $"{Name}:{message}");

    public void Receive(string message) => Inbox.Add(message);
}

public static class MediatorPattern
{
    public static IReadOnlyList<string> Demo()
    {
        var mediator = new ChatMediator();
        var alex = new Participant("alex", mediator);
        var river = new Participant("river", mediator);
        alex.Send("ready");
        return river.Inbox;
    }
}

File: java/patterns/src/main/java/org/prototypepatterns/behavioral/mediator/MediatorPattern.java
package org.prototypepatterns.behavioral.mediator;

import java.util.ArrayList;
import java.util.List;

public final class MediatorPattern {
    private MediatorPattern() {
    }

    public static final class ChatMediator {
        private final List<Participant> participants = new ArrayList<>();

        public void register(Participant participant) {
            participants.add(participant);
        }

        public void broadcast(Participant sender, String message) {
            for (Participant participant : participants) {
                if (participant != sender) {
                    participant.receive(message);
                }
            }
        }
    }

    public static final class Participant {
        private final String name;
        private final ChatMediator mediator;
        private final List<String> inbox = new ArrayList<>();

        public Participant(String name, ChatMediator mediator) {
            this.name = name;
            this.mediator = mediator;
            mediator.register(this);
        }

        public void send(String message) {
            mediator.broadcast(this, name + ":" + message);
        }

        public void receive(String message) {
            inbox.add(message);
        }

        public List<String> inbox() {
            return inbox;
        }
    }

    public static List<String> demo() {
        ChatMediator mediator = new ChatMediator();
        Participant alex = new Participant("alex", mediator);
        Participant river = new Participant("river", mediator);
        alex.send("ready");
        return river.inbox();
    }
}
